Exercici 1
Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. 
Explora les dades que conté,explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. 
Justifica la teva elecció.

Neteja el data set.

Airlines Delay: Airline on-time statistics and delay causes
Important:Redueix la dimensió del dataset de manera aleatòria per tal d'obtenir un dataset de només 200.000 registres. 
Tots els exercicis s'han de fer amb aquest dataset reduït.

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('DelayedFlights.csv')


In [3]:
df.head()

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0           0  2008      1           3          4   2003.0        1955   
1           1  2008      1           3          4    754.0         735   
2           2  2008      1           3          4    628.0         620   
3           4  2008      1           3          4   1829.0        1755   
4           5  2008      1           3          4   1940.0        1915   

   ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0   2211.0        2225            WN  ...     4.0     8.0          0   
1   1002.0        1000            WN  ...     5.0    10.0          0   
2    804.0         750            WN  ...     3.0    17.0          0   
3   1959.0        1925            WN  ...     3.0    10.0          0   
4   2121.0        2110            WN  ...     4.0    10.0          0   

   CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                 N         0           NaN           NaN      NaN   
1                 N         0           NaN           NaN      NaN   
2                 N         0           NaN           NaN      NaN   
3                 N         0           2.0           0.0      0.0   
4                 N         0           NaN           NaN      NaN   

  SecurityDelay  LateAircraftDelay  
0           NaN                NaN  
1           NaN                NaN  
2           NaN                NaN  
3           0.0               32.0  
4           NaN                NaN  

[5 rows x 30 columns]

In [4]:
df.info()  #estudiem el tipus de dades que es fa servir a cada columna

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

El fitxer compta amb 30 columnes amb diferents informacions sobre els vols. Estem interessats en estudiar els retards del vols. 
Comencem mirant quines dades són necessaris pels estudis:

-unnamed: es pot eliminar, són índex que no farem servir per l'estudi  
-year: tot l'estudi és pel 2008, es pot eliminar  
-month, DayofMonth, DayofWeek: no ho farem servir per aquest estudi
-DepTime: hora de sortida  
-CRSDepTime: hora oficial de sortida  
-ArrTime: hora d'arribada  
-CRSArrTime: hora d'arribada oficial  
-UniqueCarrier: nom de la companyia. aquest valor per aquest estudi no el farem servir amb el número serà suficient  
-FlightNum: número del vol. Útil per identificar el vol   
-TailNum: número de cua, es pot eliminar  
-ActualElapsedTime: duració real del vol  
-CRSElapsedTime: durada programada del vol  
-AirTime: temps de vol.   
-ArrDelay: delay a l'arribada. Aquesta és una variable important per poder fer l'estudi  
-DepDelay: delay a la sortida. Encara que surti tard, potser que després arribi a l'hora. No la farem servir  
-Origin: ciutat de sortida. No necessitem per fer l'estudi. Ens importa més el temps de vol independentment del lloc de sortida    
-Dest: ciutat de destí . No ho necessitem, com deiem a l'origen   
-Distance: distancia. Preferim treballar amb temps que no amb distància.    
-TaxiIn / TaxiOut : temps que triga l'avió desde que arriba a pista, fins que pujen. El mateix però en sentit contrari  
-Cancelled: si va ser o no cancellat  
-CancelledCode: tipus de cancelació. Dona diferents opcions A,B,C  
-Diverted: canvi de direcció  
-CarrierDelay: endarreriment per maletes, càrrega etc  
-WeatherDelay: endarreriment pel temps  
-NASDelay: enderreriment pel sistema nacional de seguretat  
-SecurityDelay: endarreriment per temes de seguretat  
-LateAircraftDelay: endarreriment per arribar tard l'avió  


In [5]:
np.unique(df["Year"])   #descartem la columna de year pq totes les dades són del 2008

array([2008], dtype=int64)

In [6]:
np.unique(df["UniqueCarrier"]) #nom de les companyies

array(['9E', 'AA', 'AQ', 'AS', 'B6', 'CO', 'DL', 'EV', 'F9', 'FL', 'HA',
       'MQ', 'NW', 'OH', 'OO', 'UA', 'US', 'WN', 'XE', 'YV'], dtype=object)

In [7]:
np.unique(df["CancellationCode"]) #tipus de cancelacions

array(['A', 'B', 'C', 'N'], dtype=object)

Mirem el número de null. Podem veure que tenim molts especialment en els delays.

In [8]:
df.isnull().sum()

Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64

In [9]:
df.describe().T #Mirem les dades. Fem servir transposada per veure totes les dades.   
                # Per poder fer els estudis necessitaré que les dades siguin números

count          mean           std     min        25%  \
Unnamed: 0         1936758.0  3.341651e+06  2.066065e+06     0.0  1517452.5   
Year               1936758.0  2.008000e+03  0.000000e+00  2008.0     2008.0   
Month              1936758.0  6.111106e+00  3.482546e+00     1.0        3.0   
DayofMonth         1936758.0  1.575347e+01  8.776272e+00     1.0        8.0   
DayOfWeek          1936758.0  3.984827e+00  1.995966e+00     1.0        2.0   
DepTime            1936758.0  1.518534e+03  4.504853e+02     1.0     1203.0   
CRSDepTime         1936758.0  1.467473e+03  4.247668e+02     0.0     1135.0   
ArrTime            1929648.0  1.610141e+03  5.481781e+02     1.0     1316.0   
CRSArrTime         1936758.0  1.634225e+03  4.646347e+02     0.0     1325.0   
FlightNum          1936758.0  2.184263e+03  1.944702e+03     1.0      610.0   
ActualElapsedTime  1928371.0  1.333059e+02  7.206007e+01    14.0       80.0   
CRSElapsedTime     1936560.0  1.343027e+02  7.134144e+01   -25.0       82.0   
AirTime            1928371.0  1.082771e+02  6.864261e+01     0.0       58.0   
ArrDelay           1928371.0  4.219988e+01  5.678472e+01  -109.0        9.0   
DepDelay           1936758.0  4.318518e+01  5.340250e+01     6.0       12.0   
Distance           1936758.0  7.656862e+02  5.744797e+02    11.0      338.0   
TaxiIn             1929648.0  6.812975e+00  5.273595e+00     0.0        4.0   
TaxiOut            1936303.0  1.823220e+01  1.433853e+01     0.0       10.0   
Cancelled          1936758.0  3.268348e-04  1.807562e-02     0.0        0.0   
Diverted           1936758.0  4.003598e-03  6.314722e-02     0.0        0.0   
CarrierDelay       1247488.0  1.917940e+01  4.354621e+01     0.0        0.0   
WeatherDelay       1247488.0  3.703571e+00  2.149290e+01     0.0        0.0   
NASDelay           1247488.0  1.502164e+01  3.383305e+01     0.0        0.0   
SecurityDelay      1247488.0  9.013714e-02  2.022714e+00     0.0        0.0   
LateAircraftDelay  1247488.0  2.529647e+01  4.205486e+01     0.0        0.0   

                         50%         75%        max  
Unnamed: 0         3242558.0  4972466.75  7009727.0  
Year                  2008.0     2008.00     2008.0  
Month                    6.0        9.00       12.0  
DayofMonth              16.0       23.00       31.0  
DayOfWeek                4.0        6.00        7.0  
DepTime               1545.0     1900.00     2400.0  
CRSDepTime            1510.0     1815.00     2359.0  
ArrTime               1715.0     2030.00     2400.0  
CRSArrTime            1705.0     2014.00     2400.0  
FlightNum             1543.0     3422.00     9742.0  
ActualElapsedTime      116.0      165.00     1114.0  
CRSElapsedTime         116.0      165.00      660.0  
AirTime                 90.0      137.00     1091.0  
ArrDelay                24.0       56.00     2461.0  
DepDelay                24.0       53.00     2467.0  
Distance               606.0      998.00     4962.0  
TaxiIn                   6.0        8.00      240.0  
TaxiOut                 14.0       21.00      422.0  
Cancelled                0.0        0.00        1.0  
Diverted                 0.0        0.00        1.0  
CarrierDelay             2.0       21.00     2436.0  
WeatherDelay             0.0        0.00     1352.0  
NASDelay                 2.0       15.00     1357.0  
SecurityDelay            0.0        0.00      392.0  
LateAircraftDelay        8.0       33.00     1316.0

In [10]:
df[df["ArrDelay"]<15].shape #miro els valors que no són endarreriments. 
                            #aquest valor el marco considerant trigar menys de 15minuts

(680883, 30)

In [11]:
df.Cancelled.value_counts() #miro si te relació amb cancelats.

Cancelled
0    1936125
1        633
Name: count, dtype: int64

In [12]:
df.Diverted.value_counts() #el nombre d'avions desviats

Diverted
0    1929004
1       7754
Name: count, dtype: int64

In [13]:
df.Diverted.value_counts()[1] +df.Cancelled.value_counts()[1]

8387

In [14]:
df["ArrDelay"].isnull().sum() #los null fan referència a vols desviats o cancelats

8387

In [15]:
df.ArrDelay[df["ArrDelay"]<15]=0

C:\Users\maria\AppData\Local\Temp\ipykernel_14288\2400250138.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.ArrDelay[df["ArrDelay"]<15]=0


In [16]:
#agafo els no cancellats i no desviats
df = df[(df.Cancelled == 0) & (df.Diverted == 0)]   

In [17]:
#veiem que els nan procedeixen del cancellats i desviats
#com no són objecte d'estudi es col.looc en zero.
df.isnull().sum()

Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                   0
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   3
ActualElapsedTime         0
CRSElapsedTime            0
AirTime                   0
ArrDelay                  0
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                    0
TaxiOut                   0
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         680883
WeatherDelay         680883
NASDelay             680883
SecurityDelay        680883
LateAircraftDelay    680883
dtype: int64

In [18]:
#busquem duplicats. No tenen
df.duplicated().sum()

0

In [19]:
#selecciono les columnes
#doble corchete, sobre el mismo df
df1 = df[["FlightNum","AirTime","ArrDelay","UniqueCarrier"]]
df1.head()

FlightNum  AirTime  ArrDelay UniqueCarrier
0        335    116.0       0.0            WN
1       3231    113.0       0.0            WN
2        448     76.0       0.0            WN
3       3920     77.0      34.0            WN
4        378     87.0       0.0            WN

Reduïm el dataframe amb 20000 dades

In [20]:
#reduïm amb el random state 18 per fer-ho sempre igual
df2=df1.sample(20000,random_state=18)
df2.shape


(20000, 4)

Exercici 2
Fes un informe complet del dataset:

Resumeix estadísticament el dataset i les columnes d’interès. 
Fes una anàlisi estadístic del que consideris rellevant.
Mostra quina és les aerolínies amb més endarreriments acumulats.
Quins són els vols més llargs?



Mostra quina és les aerolínies amb més endarreriments acumulats.
Aquesta pregunta pot interpretar-se com la mitjana d'endarreriments o la suma acumulada de retards.

In [27]:
df2.describe()

FlightNum       AirTime      ArrDelay
count  20000.00000  20000.000000  20000.000000
mean    2162.91660    108.462500     40.299750
std     1938.07194     68.915572     56.992786
min        1.00000      5.000000      0.000000
25%      588.00000     58.000000      0.000000
50%     1527.00000     90.000000     24.000000
75%     3410.00000    137.000000     55.000000
max     7829.00000    623.000000   1143.000000

El retard de valor mig és de uns 40minuts, amb una desviació del 57 minuts. Si veiem els quartils observem que la majoria de vols amb endarreriments estan entre 0 i 24 minuts. A la part superior trobem valors molt grans 1143minuts que fan augmentar el valor mig. Aquest són valor molt extremals.

Pel que fa al temps de vol, el rang de temps de vol és molt ampli va de 5 min - 623 min. En promig són vols de menys de 2h amb una desviació de 69 minuts. Tot i que el 50% està amb 90min.

In [22]:
#mitja del temps d'endarreriment
df2.groupby("UniqueCarrier")["ArrDelay"].mean().sort_values(ascending=False).head(10)

UniqueCarrier
YV    54.072981
B6    53.173310
XE    51.315392
OH    50.117241
EV    49.112864
OO    46.239489
AA    44.384236
UA    44.147436
FL    43.916667
HA    42.972222
Name: ArrDelay, dtype: float64

In [23]:
#suma acumulada dels endarreriments.
df2.groupby("UniqueCarrier")["ArrDelay"].sum().sort_values(ascending=False).head(10)

UniqueCarrier
WN    107876.0
AA     90100.0
UA     61983.0
OO     61591.0
MQ     60477.0
XE     54343.0
DL     43910.0
CO     40657.0
EV     40469.0
US     34956.0
Name: ArrDelay, dtype: float64

Quin és el vol més llarg?


In [30]:
temps_vol = df2.sort_values("AirTime",ascending=False)
temps_vol.head()


FlightNum  AirTime  ArrDelay UniqueCarrier
877108          15    623.0      26.0            CO
1618877         15    607.0       0.0            CO
1072562         15    605.0       0.0            CO
1520545         15    601.0       0.0            CO
362681          15    589.0       0.0            CO

Exercici 3
Exporta el dataset net i amb les noves columnes a Excel.

In [31]:
df2.to_excel("RetardsVols.xlsx")